app/src/main/java/com/example/coen268/CreateAccountActivity.java

```diff
@@ -124,6 +129,19 @@ public void onComplete(@NonNull Task<AuthResult> task) {
                         }
                     }
                 });
+        // Register a document in real time database for business reservation info
+        if (user.getAccountType().equals(Constants.ACCOUNT_TYPE_BUSINESS)) {
+            DatabaseReference databaseReservation = FirebaseDatabase.getInstance().getReference("Reservations");
```

Comment:
This section needs to be moved to within the callback function (around line 100). Otherwise this code will run even if the business already exists.

Reply:
According to your database query, createAccount() will first query the firestore datasbase and only call createAccountWithEmailPassword if the user doesn't exist in the database. When createAccountWithEmailPassword is called, it's guaranteed that the business is not in the database.

Reply:
This is for business account only. The other entry point user creation cannot possibly hit this code since it's guarded by Constants.ACCOUNT_TYPE_BUSINESS

Reply:
Yes, but the query function is asynchronous. It's possible your code will run before the query completes. You may end up with duplicate entries in your database in that case. You need to move this code into the onComplete function to guarantee that the business is unique or not and create the record accordingly.

Reply:
Isn't createAccountWithEmailPassword() called inside the completion handler (aka onComplete)? When onCompleted is asynchronously invoked, the query is done. When the query is done, calls to createAccountWithEmailPassword is synchronous

Reply:
Ah, sorry my bad, for some reason I thought this was within the createAccount function.